packages/react-graphql/src/hooks/graphql-document.ts

```diff
@@ -0,0 +1,60 @@
+import {useState, useEffect, useCallback} from 'react';
+import {OperationVariables} from 'apollo-client';
+import {DocumentNode} from 'graphql-typed';
+import {useMountedRef} from '@shopify/react-hooks';
+
+import {AsyncQueryComponentType} from '..';
+
+export default function useGraphQLDocument<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  documentOrComponent:
+    | DocumentNode<Data, Variables>
+    | AsyncQueryComponentType<Data, Variables, DeepPartial>,
+) {
+  const [document, setDocument] = useState<DocumentNode<
+    Data,
+    Variables
+  > | null>(() => {
+    if (isDocumentNode(documentOrComponent)) {
+      return documentOrComponent;
+    } else {
+      return documentOrComponent.resolved;
+    }
+  });
```

Comment:
I wonder if this is a case where we can break the rule of hooks by short circuiting here. Maybe better as an optimization later, but I think it will at least be an option.